### PR TITLE
Mongo snap only

### PR DESCRIPTION
**Pull request by wallyworld**
There's 3 main changes here, mostly big deletions:
- mongo is now only installed from the juju-db snap
- server side transactions by default
- removal of apiserver and state juju restore code

The first commit is the removal of the cd from apiserver and state which existed to implement juju restore.
A lot of this code assumed mongo 3 etc and we needed to remove it anyway, so easier to do it now since mongo 3 support is also removed. Another side effect is we no longer need to track mongo version in agent.conf. And the restore watcher worker is also obsolete.

A necessary mgo test change is that repliacset support needs to be enabled on all mgo test suites, or else tests will fail on mongo 4.x.

The install of mongo on the controllers is now snap only, and 4.x only, meaning we can also turn on server side txns.
Todo is to figure out how to enable numa support for the snap.
Also todo is to delete juju restore CLI etc

## QA steps

```sh
juju bootstrap --bootstrap-series=xenial
juju status, juju deploy etc
```

check logs for errors

```sh
juju dump-db
juju enable-ha
juju show-controller # to check replicaset status
```

log into mongo and check `rs.config()`

```sh
juju create-backup
```

repeat on a focal controller also

### CODE REVIEW

**Comment by wallyworld** [comment]
[juju uses mongo 4.0 from snap only](https://trello.com/c/5eTBpXPm/1316-juju-uses-mongo-40-from-snap-only)

File: service/snap/snap.go [hunk 1]
```diff
@@ -202,7 +202,7 @@ type Service struct {
 //
 // If no BackgroundServices are provided, Service will wrap all of the snap's
 // background services.
-func NewService(mainSnap string, serviceName string, conf common.Conf, snapPath string, Channel string, ConfinementPolicy string, backgroundServices []BackgroundService, prerequisites []App) (Service, error) {
+func NewService(mainSnap string, serviceName string, conf common.Conf, snapPath, configDir string, Channel string, ConfinementPolicy string, backgroundServices []BackgroundService, prerequisites []App) (Service, error) {
```
Comment (hpidcock) [inline service/snap/snap.go:205]: add type for consistency

File: testing/mgo.go [hunk 2]
```diff
@@ -16,11 +16,13 @@ import (
 // tests. For tests that care about security (which should be few), use
 // MgoSSLTestPackage.
 func MgoTestPackage(t *testing.T) {
-	gitjujutesting.MgoTestPackage(t, nil)
+	jujutesting.MgoServer.EnableReplicaSet = true
```
Comment (hpidcock) [inline testing/mgo.go:19]: so we always test with replica set?

**Review by hpidcock — COMMENTED** [review]
Few questions will test then approve

**Review by hpidcock — APPROVED** [review]

File: testing/mgo.go (hunk 2 above)
Comment (wallyworld) [reply testing/mgo.go:19]: yeah, mongo 4 needs it

File: service/snap/snap.go (hunk 1 above)
Comment (wallyworld) [reply service/snap/snap.go:205]: what did you mean here?

File: apiserver/facades/schema.json [hunk 3]
```diff
@@ -5124,10 +5124,6 @@
                         }
                     }
                 },
-                "FinishRestore": {
```
Comment (manadart) [inline apiserver/facades/schema.json:5127]: Shouldn't we still bump this version? Even if we delete the old implementations for a 3.0 release, it at least makes it clear that there is a new facade that no longer supports restore.

**Review by manadart — COMMENTED** [review]
Alpha space change looks fine.

File: apiserver/facades/schema.json (hunk 3 above)
Comment (wallyworld) [reply apiserver/facades/schema.json:5127]: Good idea

**Comment by wallyworld** [comment]
!!build!!

**Comment by wallyworld** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
This will break windows "unit" tests as they depend on mongo and because we've moved to snaps. We could in theory run mongo in WSL but I'm actually unsure if that's doable. Alternatively, we re-write all the tests that use the JujuConnSuite as real unit tests, but that also means the status test suite :vomiting_face:

**Comment by manadart** [comment]
Tests passed for the merge job, but we got an error on the actual merge:
```
11:50:01 Merging PR[GHPullRequest@2dc9b381[base=org.kohsuke.github.GHCommitPointer@2ec8add7,head=org.kohsuke.github.GHCommitPointer@33673dda,additions=579,commits=13,merged=false,draft=false,mergeable=true,deletions=6624,assignee=<null>,assignees={},state=open,number=12436,comments=3,labels=[],title=Mongo snap only,milestone=<null>,locked=false,url=https://api.github.com/repos/juju/juju/pulls/12436,id=535671590,nodeId=MDExOlB1bGxSZXF1ZXN0NTM1NjcxNTkw,createdAt=2020-12-10T05:38:21Z,updatedAt=2020-12-15T09:51:31Z]] is anonymous: false
11:50:01 ERROR: Step ‘Github Pull Request Merger’ aborted due to exception: 
11:50:01 java.io.FileNotFoundException: https://api.github.com/repos/juju/juju/pulls/12436/merge
11:50:01 	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1896)
11:50:01 	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1498)
11:50:01 	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
11:50:01 	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:352)
11:50:01 	at org.kohsuke.github.GitHubHttpUrlConnectionClient.getResponseInfo(GitHubHttpUrlConnectionClient.java:69)
11:50:01 	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:401)
...
```

**Comment by manadart** [comment]
$$merge$$

File: state/open.go [hunk 4]
```diff
@@ -130,18 +128,21 @@ func newState(
...
 	mongodb := session.DB(jujuDB)
-	sstxn := featureflag.Enabled(feature.MongoDbSSTXN)
+	sstxn := txn.SupportsServerSideTransactions(mongodb)
+
+	// TODO(wallyworld) - server side txns have issues on mongo 4.0
+	// (due to juju making bad txn.Ops, and potentially mongo 4 issues)
+	// Symptoms include intermittent success in applying a txn and reading results back,
+	// plus txn.Op slices that are wrong, eg insert dup or missing asserts.
+	// Related test failures here:
+	// https://jenkins.juju.canonical.com/job/github-make-check-juju/9317/testReport/
```
Comment (manadart) [inline state/open.go:138]: @wallyworld, bad news.

I investigated `FAIL: deployer_test.go:187: deployerSuite.TestDeployRecallRemoveSubordinates`. I dumped the transaction ops and DB for a failure:
https://pastebin.canonical.com/p/t3f3DWY6sN/

It looks like the DB is congruent with the assertions, so unless I am mistaken there's a deeper issue than a change in txn execution leniency.

**Comment by petevg** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**